atef/widgets/happi.py

```diff
@@ -0,0 +1,470 @@
+"""
+Widget classes designed for atef-to-happi interaction.
+"""
+
+from __future__ import annotations
+
+import logging
+from typing import Any, ClassVar, Dict, List, Optional, Union
+
+import happi
+import ophyd
+from happi.qt.model import (HappiDeviceListView, HappiDeviceTreeView,
+                            HappiViewMixin)
+from qtpy import QtCore, QtGui, QtWidgets
+from qtpy.QtWidgets import QWidget
+
+from ..qt_helpers import ThreadWorker, copy_to_clipboard
+from .core import DesignerDisplay
+from .ophyd import OphydDeviceTableWidget
+
+logger = logging.getLogger(__name__)
+
+
+class HappiSearchWidget(DesignerDisplay, QWidget):
```

Comment:
This widget looks :+1: to me, I like the implementation. In isolation this behaves pretty snappily with just a one-time startup cost for reading the happi database.

In the config gui, I'm imagining:
- instead of adding an empty line edit, the add device button would open a dialog
- this dialog would include the HappiSearchWidget and probably the metadata widget
- the confirm/close options at the bottom of the dialog would close the dialog and add an entry of the current selection on confirm

Reply:
That sounds good from the config gui perspective. I think keeping the happi search widgets + metadata widget (and explicitly not the component widget) makes sense.